wfl/calculators/aims.py

```diff
@@ -0,0 +1,165 @@
+"""
+FHI-Aims Calculator
+"""
+
+import shlex
+
+from copy import deepcopy
+import numpy as np
+
+from ase.calculators.calculator import all_changes
+from ase.calculators.aims import Aims as ASE_Aims
+try:
+    from ase.calculators.aims import AimsProfile
+except ImportError:
+    AimsProfile = None
+
+from .wfl_fileio_calculator import WFLFileIOCalculator
+from .utils import handle_nonperiodic
+
+# NOMAD compatible, see https://nomad-lab.eu/prod/rae/gui/uploads
+_default_keep_files = ["control.in", "geometry.in", "aims.out"]
+_default_properties = ["energy", "forces", "stress"]
+
+
+class Aims(WFLFileIOCalculator, ASE_Aims):
+    """Extension of ASE's Aims calculator that can be used by wfl.calculators.generic.
+
+    The ```directory``` argument from the basic calculator implementation in ASE cannot be present.
+    Use ```rundir_prefix``` and ```workdir``` instead.
+
+    Parameters
+    ----------
+    keep_files: bool / None / "default" / list(str), default "default"
+        What kind of files to keep from the run:
+            - True :        Everything kept.
+            - None, False : Nothing kept, unless calculation fails.
+            - "default" :   Only ones needed for NOMAD uploads ('control.in', 'geometry.in', 'aims.out') are kept.
+            - list(str) :   List of file globs to save.
+    rundir_prefix: str / Path, default 'run\_Aims\_'
+        Run directory name prefix.
+    workdir: str / Path, default . at calculate time
+        Path in which the run directory will be created.
+    scratchdir: str / Path, default None
+        Temporary directory to execute calculations in and delete or copy back results (set by ```keep_files```)
+        if needed.  For example, directory on a local disk with fast file I/O.
+    calculator_exec: str
+        command for Aims, without any prefix or redirection set.
+        For example: "srun -n 4 /path/to/aims.*.scalapack.mpi.x".
+        Mutually exclusive with ```command```.
+    get_k_grid: func(Atoms) -> str='<k1> <k2> <k3>', default None
```

Comment:
Is there really a need for a general function here, or is supporting something equivalent to k-spacing (which VASP and CASTEP use) sufficient? If there is a need for a general function, that almost certainly applies to other calculators, and we should provide a general-use function.

Reply:
There is an argument `k_grid_density` in FHI-Aims which might be what we would like to have. I personally never used it (always used `k_grid`). I'm trying to find out more about it. If `k_grid_density` is the FHI-Aims equivalent to the k-spacing arguments in other codes (and I think it is) I can also remove the `get_k_grid`-functionality.

Reply:
Ok, so `k_grid_density` is the keyword we can use here instead of introducing the `get_k_grid`-functionality. I'll remove it therefore.
The bad news is that the latest ASE release (3.22.1) does not support this keyword in the Aims-Calculator. (Version 3.23.0b1 does, though.)